crates/ruff_python_whitespace/src/tokenizer.rs

```diff
@@ -516,6 +510,137 @@ impl<'a> SimpleTokenizer<'a> {
     pub fn skip_trivia(self) -> impl Iterator<Item = Token> + DoubleEndedIterator + 'a {
         self.filter(|t| !t.kind().is_trivia())
     }
+
+    /// Given the position of `#` tokens on a line, test if any `#` is the start of a comment and, if so, return the
+    /// length of the comment.
+    ///
+    /// The challenge is that `#` tokens can also appear inside of strings:
+    ///
+    /// ```python
+    /// ' #not a comment'
+    /// ```
+    ///
+    /// This looks innocent but is the `'` really the start of the new string or could it be a closing delimiter
+    /// of a previously started string:
+    ///
+    /// ```python
+    /// ' a string\
+    /// ` # a comment '
+    /// ```
+    ///
+    /// The only way to reliability tell whether the `#` is a comment when the comment contains a quote char is
+    /// to forward lex all strings and comments and test if there's any unclosed string literal. If so, then
+    /// the hash cannot be a comment.
+    fn comment_from_hash_positions(&self, hash_positions: &[TextSize]) -> Option<TextSize> {
+        // Iterate over the `#` positions from the start to the end of the line.
+        // This is necessary to correctly support `a # comment # comment`.
+        for possible_start in hash_positions.iter().rev() {
```

Comment:
The case where we have multiple hashes could probably be sped up to avoid finding the unterminated string repeatedly because I think this is extremely rare.